### PR TITLE
Set default view configuration in datasource properties json

### LINKED ISSUE #4298: Set initial layer color trough datasource-properties.json
It would be nice to set the color of a layer in the `datasource-properties.json`. I still think the users should be able to edit that color, but this could be the default.

### LINKED ISSUE #4378: Remove segmentationOpacity from DataSet
Instead use the `alpha` value of the `SegmentationLayer` 

This is a followup of #4357

### PR DESCRIPTION
**Pull request by youri-k**
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
for the evolutions:
- after the evolution the segmentationOpacity of the configuration element should be gone
- instead the old segmentationOpacity value should be the alpha value of a newly created entry in the layers dict with the name of the related segmentation layer as key
- the reverse evolution should take the alpha value from segmentation layer entry in the layers dict and save it as segmentationOpacity
- to confirm that the configuration entries are still correct, open the dataset edit view and look at the view configurations

for the disk defaults:
- if a view configuration is supplied in the datasource properties json, it should take precedence over the normal default, but if a existing dataset config takes precedence over the disk default
- e.g. change the `alpha` value of a color layer (default is 100) 

### Issues:
- fixes #4298 
- fixes #4378 

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [x] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable
- ~~[ ] Updated [documentation](../blob/master/docs) if applicable~~
- ~~[ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes~~
- [x] Needs datastore update after deployment
- [x] Ready for review

### CODE REVIEW

**Comment by youri-k** [comment]
@MichaelBuessemeyer this pr should also take care of #4378. Could you do the necessary frontend part? This involves the the dataset edit view and the rendering in the tracing view. Thank you :slightly_smiling_face:

**Comment by MichaelBuessemeyer** [comment]
@youri-k If we want to 
> Instead use the alpha value of the SegmentationLayer

we need to add the configs of the segmentation layer to the other layer configs in the DatasetConfiguration under the property "layers". Currently, that's not the case and will be needed.

Can you please do the backend part of this and write a conversion script, that removes the "segmentationOpacity" from DatasetConfiguration and adds a layer configuration for the segmentation to the `layers` property of DatasetConfiguration.

**Comment by MichaelBuessemeyer** [comment]
@philippotto Could you please check out this weird flow error I cannot fix? Thanks a lot :smile:

**Comment by youri-k** [comment]
@fm3 can you review the backend part again, especially the evolutions :slightly_smiling_face:

File: frontend/javascripts/oxalis/model/sagas/volumetracing_saga.js [hunk 1]
```diff
@@ -55,10 +55,13 @@ function* warnOfTooLowOpacity(): Saga<void> {
   if (yield* select(state => state.tracing.volume == null)) {
     return;
   }
-
-  const isOpacityTooLow = yield* select(
-    state => state.datasetConfiguration.segmentationOpacity < 10,
-  );
+  const segmentationLayerName = yield* call([Model, Model.getSegmentationLayerName]);
```
Comment (philippotto) [inline frontend/javascripts/oxalis/model/sagas/volumetracing_saga.js]: I think, you can simply `return` if `segmentationLayerName` is null. This should never happen anyway (since we already know there is a volume tracing when we hit this code path).

File: frontend/javascripts/oxalis/view/settings/dataset_settings_view.js [hunk 2]
```diff
@@ -265,50 +265,45 @@ class DatasetSettings extends React.PureComponent<DatasetSettingsProps, State> {
 
   getLayerSettings = (
     layerName: string,
-    layer: ?DatasetLayerConfiguration,
+    layerConfiguration: ?DatasetLayerConfiguration,
     isColorLayer: boolean = true,
   ) => {
-    // Ensure that a color layer needs a layer.
-    if (isColorLayer && !layer) {
+    // Ensure that a every layer needs a layer configuration.
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/settings/dataset_settings_view.js]:
```suggestion
    // Ensure that every layer needs a layer configuration.
```

File: frontend/javascripts/oxalis/view/settings/dataset_settings_view.js [hunk 3]
```diff
@@ -265,50 +265,45 @@ class DatasetSettings extends React.PureComponent<DatasetSettingsProps, State> {
...
+    if (!layerConfiguration) {
       return null;
     }
     const elementClass = getElementClass(this.props.dataset, layerName);
     const isDisabled =
-      isColorLayer && layer != null
-        ? layer.isDisabled
+      isColorLayer && layerConfiguration != null
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/settings/dataset_settings_view.js]: Why is this check necessary? In line 272 you already return from this function if the var is null?

File: frontend/javascripts/oxalis/view/settings/dataset_settings_view.js [hunk 4]
```diff
@@ -265,50 +265,45 @@ class DatasetSettings extends React.PureComponent<DatasetSettingsProps, State> {
...
-              onChange={
-                isColorLayer
-                  ? _.partial(this.props.onChangeLayer, layerName, "alpha")
-                  : _.partial(this.props.onChange, "segmentationOpacity")
-              }
+              value={layerConfiguration.alpha}
+              onChange={_.partial(this.props.onChangeLayer, layerName, "alpha")}
             />
-            {isColorLayer && layer != null ? (
+            {isColorLayer && layerConfiguration != null ? (
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/settings/dataset_settings_view.js]: Same here and in line 288.

File: frontend/javascripts/oxalis/api/api_latest.js [hunk 5]
```diff
@@ -1040,8 +1043,26 @@ class DataApi {
    * @example
    * const segmentationOpacity = api.data.getConfiguration("segmentationOpacity");
    */
-  getConfiguration(key: $Keys<DatasetConfiguration>) {
-    return Store.getState().datasetConfiguration[key];
+  getConfiguration(key: $Keys<DatasetConfiguration> | OutdatedDatasetConfigurationKeys) {
+    switch (key) {
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply frontend/javascripts/oxalis/api/api_latest.js]: I think I wouldn't do it right now, because the change is relatively minor and from what I just checked there are still api scripts in very active use that request API version 2. Bumping the latest version now would make those fail. We should push people (again) to use the newest API version before doing that out of the blue.
I think the specific deprecation warnings are a good thing, so I would include them in any case. Script writers will see more easily what they need to migrate :)

File: frontend/javascripts/oxalis/view/settings/dataset_settings_view.js (hunk 3 above)
Comment (MichaelBuessemeyer) [reply frontend/javascripts/oxalis/view/settings/dataset_settings_view.js]: Nice thanks for the catch :eyes:

File: frontend/javascripts/oxalis/api/api_latest.js [hunk 6]
```diff
@@ -693,9 +696,9 @@ class DataApi {
    */
   getVolumeTracingLayerName(): string {
     // TODO: Rename method to getSegmentationLayerName() and increase api version
-    const segmentationLayer = this.model.getSegmentationLayer();
-    assertExists(segmentationLayer, "Segmentation layer not found!");
-    return segmentationLayer.name;
+    const segmentationLayerName = this.model.getSegmentationLayerName();
+    assertExists(segmentationLayerName, "Segmentation layer not found!");
+    return segmentationLayerName || "";
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/oxalis/api/api_latest.js]: I need the alternative of  `""` here as flow does not know that at this point `segmentationLayerName` is not null.

Is this fine how I handled this or should I revert this change and use the old version.

This change here is just some refactoring that came into my mind.

Comment (philippotto) [reply frontend/javascripts/oxalis/api/api_latest.js]: I'd replace `assertExists` with a normal if-null-then-throw logic. That way, flow will understand that the following code can safely access `segmentationLayerName`.

```
if (!segmentationLayerName) {
  throw new Exception("Segmentation layer not found!");
}
return segmentationLayerName;
```

**Comment by MichaelBuessemeyer** [comment]
@philippotto As a last thing. Could you please check my deprecation warning message? I think it is quite long but I do not know how to shorten it further :confused:

File: frontend/javascripts/oxalis/api/api_latest.js [hunk 7]
```diff
@@ -1051,8 +1077,32 @@ class DataApi {
...
    * api.data.setConfiguration("segmentationOpacity", 20);
    */
-  setConfiguration(key: $Keys<DatasetConfiguration>, value) {
-    Store.dispatch(updateDatasetSettingAction(key, value));
+  setConfiguration(key: $Keys<DatasetConfiguration> | OutdatedDatasetConfigurationKeys, value) {
+    console.warn(
+      `The properties segmentationOpacity and isSegmentationDisabled are no longer directly part of the data configuration. To still set 
+      them use the segmentation "layers" configuration. This object has settings for the segmentation layer with the properties "alpha" 
+      and "isDisabled". Use "const layerSettings = api.data.getConfiguration('layers'); layerSettings[<segmentationLayerName>].alpha = 40;
+      api.data.setConfiguration('layers', layerSettings);" to set these properties.`,
```
Comment (philippotto) [inline frontend/javascripts/oxalis/api/api_latest.js]: I think, you can make it a bit shorter by omitting the part which explains the code, as that should be self-explanatory:
```
The properties segmentationOpacity and isSegmentationDisabled are no longer directly part of the data configuration. Instead, they are part of the segmentation layer configuration and can be set as follows:
const layerSettings = api.data.getConfiguration('layers');
layerSettings[segmentationLayerName].alpha = 40;
layerSettings[segmentationLayerName].isDisabled = false;
api.data.setConfiguration('layers', layerSettings);"
```

**Comment by MichaelBuessemeyer** [comment]
> we need to add the configs of the segmentation layer to the other layer configs in the DatasetConfiguration under the property "layers". Currently, that's not the case and will be needed.
> Can you please do the backend part of this and write a conversion script, that removes the "segmentationOpacity" from DatasetConfiguration and adds a layer configuration for the segmentation to the `layers` property of DatasetConfiguration.

@youri-k did you do this?

**Comment by youri-k** [comment]
Yes, it's done

**Comment by fm3** [comment]
I tested running the evolution and then switching to this branch. I now get Invalid Schema warnings in the View Configuration tab of the edit dataset page.
the json in the field shows
```
{
  "color": {
    "alpha": 50,
    "color": [
      255,
      255,
      255
    ],
    "intensityRange": [
      0,
      255
    ]
  },
  "segmentation": {
    "alpha": 20,
    "intensityRange": [
      0,
      255
    ]
  }
}
```
and the message below says 
Invalid schema: instance.segmentation requires property "color"
is this a bug in this PR or did I do something wrong?

**Comment by MichaelBuessemeyer** [comment]
@fm3 Thanks for finding this error.
 > we need to add the configs of the segmentation layer to the other layer configs in the DatasetConfiguration under the property "layers". Currently, that's not the case and will be needed.
> Can you please do the backend part of this and write a conversion script, that removes the "segmentationOpacity" from DatasetConfiguration and adds a layer configuration for the segmentation to the `layers` property of DatasetConfiguration. 

Did you execute the migration script youri wrote? This should fix the error

**Comment by MichaelBuessemeyer** [comment]
@fm3 We decided to make the color property not mandatory. The error should be resolved by now.

**Comment by MichaelBuessemeyer** [comment]
@philippotto Does the frontend part look fine too you? Could you please check this again?

**Comment by MichaelBuessemeyer** [comment]
@philippotto Could you please review the newest version of the frontend part of this pr?

**Comment by MichaelBuessemeyer** [comment]
@philippotto First of all, I am sorry that I missed merging this PR earlier.

Because I missed this, your newest PR reworking the rendering to only display the enabled layers when the GPU does not support rendering all layers created some "hidden" merge conflicts.

I tried resolving them and it looks fine now and everything works as expected on my local machine. Nevertheless could you please check the newest changes again and whether everything works on your machine? But I can ensure that the bug @fm3 found is resolved and does not appear in the current version of this PR.

**Comment by MichaelBuessemeyer** [comment]
@philippotto Thanks for finding this bug. I debugged it a bit further and it looks like the backend sends false information:
 - When I open the link of yours in a private tab and look at the network traffic the `https://viewdefaultsfromdisk.webknossos.xyz/api/dataSetConfigurations/sample_organization/2017-05-31_mSEM_scMS109_bk_100um_v01-aniso` call returns:
```
{
    "quality": 0,
    "layers": {
        "color": {
            "brightness": 0,
            "contrast": 1,
            "color": [255, 255, 255],
            "alpha": 100
        }
    },
    "highlightHoveredCellId": true,
    "fourBit": false,
    "interpolation": true,
    "renderMissingDataBlack": true
}
```
In my opinion the answer should have an alpha value of 80 and the color you set in the settings, shouldn't it?

@youri-k Could you please have a look at this issue/bug? I think it has something to do with the backend.

**Comment by youri-k** [comment]
@MichaelBuessemeyer I found the bug and fixed it :slightly_smiling_face: 
Additionally, I found something interesting: When you open a new dataset, you get the current dataset default configuration, and instantly respond to the server with this default config as your new user config. When the default config changes afterwards, you won't notice the change because your user config takes precedence. I don't know if this intended, but it's the same on the master. My feeling would be to only save a new user config, if the user actually changes his dataset settings and not just copying the default config at the point of first opening the dataset.
/cc @philippotto @fm3

**Comment by MichaelBuessemeyer** [comment]
@youri-k Awesome :D
Thanks for fixing the bug.
According to your thoughts, I think this is not an issue to discuss in this PR as this is how the behaviour has being for a long time. Could you please open a new issue for this?

**Comment by philippotto** [comment]
Cool, then we can merge this tomorrow, right? @youri-k Can you take care of this (plus migration)?

**Comment by normanrz** [comment]
Nice! Now we need to put this into voxelytics (and maybe wkcuber)
